flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/WaitingForResourcesTest.java

```diff
@@ -0,0 +1,328 @@
+
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.runtime.executiongraph.ArchivedExecutionGraph;
+import org.apache.flink.runtime.executiongraph.ErrorInfo;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.TestingExecutionGraphBuilder;
+import org.apache.flink.runtime.jobmaster.slotpool.ResourceCounter;
+import org.apache.flink.runtime.rest.handler.legacy.utils.ArchivedExecutionGraphBuilder;
+import org.apache.flink.runtime.scheduler.ExecutionGraphHandler;
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.nullValue;
+
+/** Tests for the WaitingForResources state. */
+public class WaitingForResourcesTest extends TestLogger {
+
+    /** WaitingForResources is transitioning to Executing if there are enough resources */
+    @Test
+    public void testTransitionToExecuting() {
+        StateTestContext testContext =
+                new StateTestContext(ctx -> ctx.setHasEnoughResources(() -> true));
+
+        testContext.assertStateTransitionTo(Executing.class);
+    }
+
+    @Test
+    public void testTransitionToFinishedOnFailure() {
+        StateTestContext testContext =
+                new StateTestContext(
+                        ctx -> {
+                            ctx.setHasEnoughResources(() -> true);
+                            ctx.setCreateExecutionGraphWithAvailableResources(
+                                    () -> {
+                                        throw new RuntimeException("Test");
+                                    });
+                        });
+
+        testContext.assertStateTransitionTo(Finished.class);
+    }
+
+    @Test
+    public void testDelayedResourceAvailability() {
+        StateTestContext testContext =
+                new StateTestContext(
+                        ctx -> {
+                            ctx.setHasEnoughResources(() -> false);
+                        });
+
+        // no resources --> no state transition
+        testContext.getWaitingForResources().notifyNewResourcesAvailable();
+        testContext.assertNoStateTransition();
+
+        // make resources available
+        testContext.getMockContext().setHasEnoughResources(() -> true);
+        testContext.getWaitingForResources().notifyNewResourcesAvailable();
+
+        testContext.assertStateTransitionTo(Executing.class);
+    }
```

Comment:
This test can be split into the two cases

1. not enough resources
2. enough resources